File: examples/openplatform-java-demo/src/main/java/com/enio/openplatform/crypto/RsaUtil.java
package com.enio.openplatform.crypto;

import javax.crypto.Cipher;
import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.util.Base64;

/**
 * RSA/ECB/PKCS1Padding，与 Go 服务端及常见 Java RSAUtil 对齐。
 */
public final class RsaUtil {

    private static final String RSA = "RSA/ECB/PKCS1Padding";

    private RsaUtil() {
    }

    /** 从管理端下发的 AppSecret（PKCS#8 DER Base64）加载私钥。 */
    public static PrivateKey loadPrivateKeyFromDerBase64(String appSecretBase64) throws Exception {
        byte[] der = Base64.getDecoder().decode(appSecretBase64.trim());
        PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(der);
        return KeyFactory.getInstance("RSA").generatePrivate(spec);
    }

    /** 私钥加密（密钥交换时加密 clientRandom）。 */
    public static String encryptByPrivateKey(PrivateKey privateKey, String plain) throws Exception {
        Cipher cipher = Cipher.getInstance(RSA);
        cipher.init(Cipher.ENCRYPT_MODE, privateKey);
        byte[] encrypted = cipher.doFinal(plain.getBytes(StandardCharsets.UTF_8));
        return Base64.getEncoder().encodeToString(encrypted);
    }

    /** 私钥解密（解密服务端返回的 serverPart）。 */
    public static String decryptByPrivateKey(PrivateKey privateKey, String cipherBase64) throws Exception {
        Cipher cipher = Cipher.getInstance(RSA);
        cipher.init(Cipher.DECRYPT_MODE, privateKey);
        byte[] raw = Base64.getDecoder().decode(cipherBase64);
        byte[] plain = cipher.doFinal(raw);
        return new String(plain, StandardCharsets.UTF_8);
    }
}

File: examples/openplatform-python-demo/openplatform/crypto/rsa_util.py
"""RSA/ECB/PKCS1Padding，与 Go/Java 服务端对齐。"""

from __future__ import annotations

import base64

from Crypto.Cipher import PKCS1_v1_5
from Crypto.PublicKey import RSA


def load_private_key_from_der_base64(app_secret_base64: str) -> RSA.RsaKey:
    der = base64.b64decode(app_secret_base64.strip())
    return RSA.import_key(der)


def encrypt_by_private_key(private_key: RSA.RsaKey, plain: str) -> str:
    """私钥加密（密钥交换时加密 clientRandom）。"""
    encrypted = _private_encrypt_pkcs1_type1(private_key, plain.encode("utf-8"))
    return base64.b64encode(encrypted).decode("ascii")


def decrypt_by_private_key(private_key: RSA.RsaKey, cipher_base64: str) -> str:
    """私钥解密（解密服务端公钥加密的 serverPart）。"""
    raw = base64.b64decode(cipher_base64)
    cipher = PKCS1_v1_5.new(private_key)
    plain = cipher.decrypt(raw, sentinel=None)
    if plain is None:
        raise ValueError("RSA decrypt failed")
    return plain.decode("utf-8")


def _private_encrypt_pkcs1_type1(key: RSA.RsaKey, message: bytes) -> bytes:
    k = key.size_in_bytes()
    if len(message) > k - 11:
        raise ValueError("message too long")
    padding_len = k - len(message) - 3
    em = b"\x00\x01" + (b"\xff" * padding_len) + b"\x00" + message
    m = pow(int.from_bytes(em, "big"), key.d, key.n)
    return m.to_bytes(k, "big")
